### ISSUE TITLE
Add userscripts support to Android version

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/AdguardForAndroid#34]
* [ ] Extensions settings UI (check how it looks like in Win version)
* [ ] Adding/updating userscripts

**Comment by TPS** [comment]
Just to confirm, this will be userscripts _subscription_, not just static userscripts, right?  Or both, just like user filters will be implemented?

**Comment by ameshkov** [comment]
No, filter subscription is different thing, it will be covered in version 2.3.

**Comment by TPS** [comment]
Nono, I mean, like the filter subscriptions that will be covered in v2.3, I'm hoping the userscripts' enhancement will include userscripts subscriptions as well, not just static userscripts.

**Comment by ameshkov** [comment]
Ah, got it. Yeah, there will be a possibility to add custom userscripts by url, not just use some pre-defined.

**Comment by Nick-White** [comment]
+1 I hope you will be able to specify custom userscripts from files too (not just public URLs) because I have tons of them.

**Comment by ameshkov** [comment]
@Nick-White of course you will:)

**Comment by Bluscream** [comment]
I totally second this :)

**Comment by Bluscream** [comment]
How far are we away from the 4.0? :'(

**Comment by ameshkov** [comment]
Pretty far:(

Btw, which exact userscripts do you want to have in Android?

**Comment by Bluscream** [comment]
Mostly AAK and ads.fly bypasser and some other stuff like Github Dark Script

**Comment by ameshkov** [comment]
AAK looks abandoned, though.

**Comment by Bluscream** [comment]
It doesn't to me. Whenever i wrote a issue it was resolved within 3-4 days :)

**Comment by TPS** [comment]
@Bluscream I think that @ameshkov means that [last commit was July 6, 2016](https://github.com/reek/anti-adblock-killer/commits/master).  Even so, @ameshkov, this is fundamental for AAK's functionality.

Hey, @reek, are you still with us? 😟

**Comment by TPS** [comment]
@ameshkov @Bluscream [@Reek's back](https://github.com/reek/anti-adblock-killer/issues/2359#issuecomment-259325481), so this is as important as ever. :smiley:

**Comment by ameshkov** [comment]
This is good news, thank you!:)

**Comment by TPS** [comment]
2 more relatively popular projects that need this are https://github.com/adsbypasser/adsbypasser & https://github.com/handyuserscripts/antiadware.

**Comment by reek** [comment]
Hi guys

I'm back :-)

**Comment by TPS** [comment]
Welcome back, friend! ☺

**Comment by Bluscream** [comment]
@ameshkov give this guy a beta license https://github.com/reek/anti-adblock-killer/issues/1883#issuecomment-260139899 xP

**Comment by TPS** [comment]
@ameshkov I agree w/ @Bluscream: @Reek, as a maintainer of a (rather important ;) builtin list needs full access to wherever his list might be used, so ß license might useful to him across the platforms… unless @AdGuardTeam would like to try to fix/workaround in-house. :sweat:

**Comment by Bluscream** [comment]
Any news on this ;D

**Comment by TPS** [comment]
@Bluscream "this" which?

The main issue is [milestoned for v4.0](https://github.com/AdguardTeam/AdguardForAndroid/milestone/6), so I don't expect anything on that front yet.

Though I'd like to know if @reek got his ß access yet.  (He's 👍ed above, too, BTW.)

**Comment by ameshkov** [comment]
> Though I'd like to know if @reek got his ß access yet. (He's 👍ed above, too, BTW.)

Not yet, but I am ready to provide it at any time. @reek if you need access to any AG premium product, just email me to am @ adguard.com

**Comment by TPS** [comment]
Just a thought:  Shouldn't this also allow us to do URL conversions of other types, e.g.,  http://www.Amazon.com/ → https://smile.Amazon.com/ or more advanced sorcery à la [HTTPS Everywhere](https://github.com/EFForg/https-everywhere/tree/master/src/chrome/content/rules)?

**Comment by TPS** [comment]
Another project that'd benefit from this is @jspenguin2017's [AdBlockProtector](https://github.com/jspenguin2017/AdBlockProtector), based on @reek's work, & consists of a blockscript & a filterset (which I request @AdguardTeam consider for inclusion).

**Comment by jspenguin2017** [comment]
@TPS Hi, thanks for the hard work of everyone here 😄 
But unfortunately, we are currently dropping support of all proprietary software, step by step, except Windows. My project may not stay as Userscript forever, or it may become an Userscript with custom extended API.

**Comment by TPS** [comment]
@jspenguin2017 Will *filterlist* remain publicly accessible, @ least?

**Comment by jspenguin2017** [comment]
@TPS Everything will stay open source and publicly accessible, there is no worries about that 😄

**Comment by TPS** [comment]
*Un*related to AdBlockProtector, https://greasyfork.org/en/scripts/19993-ru-adlist-js-fixes/code (from https://github.com/gorhill/uBlock/issues/2609#issuecomment-301318201) seems like a fascinating thing to try, as well.

**Comment by TPS** [comment]
> [Just a thought: Shouldn't this also allow us to do URL conversions of other types, e.g., http://www.Amazon.com/ → https://smile.Amazon.com/ or more advanced sorcery à la HTTPS Everywhere?](https://github.com/AdguardTeam/AdguardForAndroid/issues/34#issuecomment-290948397)

@ameshkov Could [`#%#(function`](https://github.com/AdguardTeam/AdguardForAndroid/issues/1436#issuecomment-328379256) be used to effect this?  Are there any docs about it &/or is this just straight (unsafe⁉️) JS?

**Comment by trmdi** [comment]
Hi, any good news about this very cool feature?

**Comment by Bluscream** [comment]
Nothing so far sadly, but the UI rehaul had priority.

**Comment by ameshkov** [comment]
> @ameshkov Could #%#(function be used to effect this? Are there any docs about it &/or is this just straight (unsafe⁉️) JS?

It is straight JS and that's why we limit those rules to AG filters only and not accept them from any third-party filters.

**Comment by ameshkov** [comment]
> Hi, any good news about this very cool feature?

Good news is that we're in process of transition to the single cross-platform filtering engine (codename CoreLibs). AG for Android v2.10 is already partly transitioned, but there's a lot more to do. Once we finish with all the products transition, they will share common functionality.

Regarding the userscripts support, it's not enough to just execute a JS script as script hosts provide an API. 

Off-topic: it becomes even more complicated given the changes in userscripts API brought by Greasemonkey 4: https://www.greasespot.net/2017/09/greasemonkey-4-for-script-authors.html
Basically, he decided to implement completely new API that is **not** backwards-compatible.

**Comment by TPS** [comment]
@ameshkov Still, `#%#(function` gives us a little something to play with in the user _filters_ while waiting.

Any1 want to join me experimenting?  The example I link to above is a rough-&-dirty implementation of #1436.  It exhibits some JS injection to potentially block reading a hardware sensor. :bow:

The simplest userscript ***I***'d like is the www.Amazon.com → smile.Amazon.com conversion, but, [looking through the source of several addons/extensions](https://github.com/search?q=smile+Amazon&type=Repositories&utf8=%E2%9C%93), they involve `window.location`, `tab.url`, or `chrome.webrequest`. Whaddy'all think?  Could these be converted successfully?

**Comment by ameshkov** [comment]
They're mostly Chrome extensions and not userscripts so direct conversion's not possible :(

**Comment by TPS** [comment]
Ok, how about [all these](https://www.google.com/?q=userscript+%22smile.amazon.com%22)?  They mostly use `window.location.*` & `location.href.*`.  Some of these could possibly be converted?  Would any of these rewrite any _initial_ requests before loading a Non-Smile www.Amazon page?

**Comment by ameshkov** [comment]
@TPS something like that for instance?

`amazon.com#%#if(window.location.host === 'amazon.com' || window.location.host === 'www.amazon.com') { window.location.host = 'smile.amazon.com'; }`

Don't forget to enable HTTPS filtering on `amazon.com`

**Comment by Bluscream** [comment]
![screenshot_20171022-171532](https://user-images.githubusercontent.com/3318223/31863335-aa46bad0-b74c-11e7-8ad1-a918b1e2d05f.png)

There needs to be a better way to find something here than scrolling and scrolling and scrolling

**Comment by ministr91** [comment]
какие новости?

**Comment by byaka** [comment]
yep, any news about this?

**Comment by vozersky** [comment]
we might think about it after we finish Core Libs integration for all the platforms

**Comment by byaka** [comment]
looks like you thinked and denied? maybe any comments?

**Comment by vbagirov** [comment]
@byaka the latest AdGuard for Android beta already supports custom userscripts, soon enough it'll be added to release version.

**Comment by byaka** [comment]
@vbagirov ah, awesome, thx!

**Comment by driversti** [comment]
Guys, I'm so happy to get the version this morning! This feature made my day! Thanks a lot to all of you :heart:
